opcua/server/history_sql.py

```diff
@@ -0,0 +1,168 @@
+        except sqlite3.Error as e:
+            print(node_id, 'Historizing SQL Delete Old Data Error:', e)
+
+        self._conn.commit()
+
+    def read_node_history(self, node, start, end, nb_values):
+        _c_read = self._conn.cursor()
+
+        if end is None:
+            end = datetime.now() + timedelta(days=1)
+        if start is None:
+            start = ua.DateTimeMinValue
+
+        node_id = self._get_table_name(node)
+
+        cont = None
+        results = []
+
+        start_time = start.isoformat(' ')
+        end_time = end.isoformat(' ')
+
+        # select values from the database
+        try:
+            for row in _c_read.execute('SELECT * FROM "{tn}" WHERE "ServerTimestamp" BETWEEN ? AND ? '
+                                       'LIMIT ?'.format(tn=node_id), (start_time, end_time, nb_values,)):
+
+                value = self._unpack_value(row[4], row[5])
+
+                dv = ua.DataValue(ua.Variant(value, ua.VariantType(row[4])))
+                dv.ServerTimestamp = row[0]
+                dv.SourceTimestamp = row[1]
+                dv.StatusCode = ua.StatusCode(row[2])
+
+                results.append(dv)
+
+        except sqlite3.Error as e:
+            print(node_id, 'Historizing SQL Read Error:', e)
+
+        return results, cont
+
+    def new_historized_event(self, event, period):
+        raise NotImplementedError
+
+    def save_event(self, event):
+        raise NotImplementedError
+
+    def read_event_history(self, start, end, evfilter):
+        raise NotImplementedError
+
+    def _pack_value(self, datavalue):
+        variant_code = datavalue.Value.VariantType.value
+        return struct.pack(self._get_pack_type(variant_code), datavalue.Value.Value)
+
+    def _unpack_value(self, variant_code, binary,):
+        return struct.unpack(self._get_pack_type(variant_code), binary)
+
+    def _get_pack_type(self, variant_code):
+        # see object_ids.py
+        if variant_code is 1:  # Bool
+            return '?'
+        elif variant_code is 2:  # Char (string with length of one in python)
+            return 'c'
+        elif variant_code is 3:  # Byte (Signed Char in python)
+            return 'b'
+        if variant_code is 4:  # Int16
+            return 'h'
+        elif variant_code is 5:  # UInt16
+            return 'H'
+        elif variant_code is 6:  # Int32
+            return 'i'
+        elif variant_code is 7:  # UInt32
+            return 'I'
+        elif variant_code is 8:  # Int64
+            return 'q'
+        elif variant_code is 9:  # UInt64
+            return 'Q'
+        elif variant_code is 10:  # Float
+            return 'f'
+        elif variant_code is 11:  # Double
+            return 'd'
+        elif variant_code in (12,):  # String
+            return "s"
+        else:
+            # FIXME: Should raise exception here that historizing of node type isn't supported in SQL storage interface
+            return None
```

Comment:
You should here check the constants not the hard-coded number, I would write e.g.:

```
...
variant_type is VariantType.Boolean:
     return '?'
...
```

By the way, there is solution to do this with [dictionary](https://bytebaker.com/2008/11/03/switch-case-statement-in-python/) which could be possibly cleaner since you have constants... @oroulet: which way do you prefere?

I have no proposal for your problem sorry....